### Overview ###

A small vet clinic is treating a virus that turns good dogs into bad dogs. There is also a dog flu circulating at the same time.

Data entry for the patients has been rushed and inconsistencies need to be addressed. The data also needs to be in the correct type for upload to the parent company's standards.

Measurements need to be converted to metric, and a new field needs to be added to flag dogs that come in with a fever.

In [1]:
# loading data

dog_data = [['Name', 'Breed', 'Good Doggy?', 'Length [in]', 'Weight Range', 'Body Temp [F]'],
    ["Buddy", "Golden Retriever", "Good", 24, 'Overweight', 101.5],
    ["Cooper", "Doberman Pinscher", "Bad", 28, 'Normal',  105.2],
    ["Max", "GERMAN Shepherd", "Good", 26, 'Normal', 100.2],
    ["Bella", "Chow Chow", "Bad", 20, 'Normal', 100.5],
    ["Lucy", " Yorkshire Terrier", "Good", 9, 'Normal', 102.0],
    ["Charlie", "Beagle", "Good", 13, 'Overweight', 100.4],
    ["Daisy", "Akita", "Bad", 24, 'Normal',99.8],
    ["Bailey", "JACK russell terrier", "Bad", 10, 'Underweight', 101.2],
    ["Lola", "Rottweiler   ", "Good", 27, 'Normal', 102.5],
    ["Sadie", "Siberian_Husky", "Bad", 23, 'Normal', 99.5],
    ["Penny", "STANDARD Poodle", "Good", 18, 'Underweight', 100.7],
    ["Bruno", "Bulldog", "Bad", 14, 'Normal', 104.1],
    ["Rosie", "Beagle", "Good", 13, 'Normal', 100.6],
    ["Duke", "    DOBERMAN PINSCHER", "Good", 28,  'Normal',100.1],
    ["Luna", "Shiba Inu", "Good", 17, 'Obese',  100.3],
    ["Scout", "Golden_Retriever", "Good", 22, 'Normal', 103.5],
    ["Rex", "Rottweiler", "Bad", 23, 'Normal', 100.2],
]

#### 1) Write a clean_string function that removes leading/trailing spaces, replaces underscores with spaces, and capitalizes the first letter of each word. Use this function to clean the 'Breed' column. ####

In [2]:
def clean_string(text):
    return text.strip().replace("_", " ").title()

for row in dog_data[1::]:
    row[1] = clean_string(row[1])

dog_data


[['Name',
  'Breed',
  'Good Doggy?',
  'Length [in]',
  'Weight Range',
  'Body Temp [F]'],
 ['Buddy', 'Golden Retriever', 'Good', 24, 'Overweight', 101.5],
 ['Cooper', 'Doberman Pinscher', 'Bad', 28, 'Normal', 105.2],
 ['Max', 'German Shepherd', 'Good', 26, 'Normal', 100.2],
 ['Bella', 'Chow Chow', 'Bad', 20, 'Normal', 100.5],
 ['Lucy', 'Yorkshire Terrier', 'Good', 9, 'Normal', 102.0],
 ['Charlie', 'Beagle', 'Good', 13, 'Overweight', 100.4],
 ['Daisy', 'Akita', 'Bad', 24, 'Normal', 99.8],
 ['Bailey', 'Jack Russell Terrier', 'Bad', 10, 'Underweight', 101.2],
 ['Lola', 'Rottweiler', 'Good', 27, 'Normal', 102.5],
 ['Sadie', 'Siberian Husky', 'Bad', 23, 'Normal', 99.5],
 ['Penny', 'Standard Poodle', 'Good', 18, 'Underweight', 100.7],
 ['Bruno', 'Bulldog', 'Bad', 14, 'Normal', 104.1],
 ['Rosie', 'Beagle', 'Good', 13, 'Normal', 100.6],
 ['Duke', 'Doberman Pinscher', 'Good', 28, 'Normal', 100.1],
 ['Luna', 'Shiba Inu', 'Good', 17, 'Obese', 100.3],
 ['Scout', 'Golden Retriever', 'Good', 22, 

#### 2) Transform lengths from inches to centimeters. ####

In [3]:
# 1 in = 2.54 cm

for i, row in enumerate(dog_data):
    if i > 0:
        row[3] = row[3] * 2.54

dog_data[0][3] = 'Length [cm]'

dog_data


[['Name',
  'Breed',
  'Good Doggy?',
  'Length [cm]',
  'Weight Range',
  'Body Temp [F]'],
 ['Buddy', 'Golden Retriever', 'Good', 60.96, 'Overweight', 101.5],
 ['Cooper', 'Doberman Pinscher', 'Bad', 71.12, 'Normal', 105.2],
 ['Max', 'German Shepherd', 'Good', 66.04, 'Normal', 100.2],
 ['Bella', 'Chow Chow', 'Bad', 50.8, 'Normal', 100.5],
 ['Lucy', 'Yorkshire Terrier', 'Good', 22.86, 'Normal', 102.0],
 ['Charlie', 'Beagle', 'Good', 33.02, 'Overweight', 100.4],
 ['Daisy', 'Akita', 'Bad', 60.96, 'Normal', 99.8],
 ['Bailey', 'Jack Russell Terrier', 'Bad', 25.4, 'Underweight', 101.2],
 ['Lola', 'Rottweiler', 'Good', 68.58, 'Normal', 102.5],
 ['Sadie', 'Siberian Husky', 'Bad', 58.42, 'Normal', 99.5],
 ['Penny', 'Standard Poodle', 'Good', 45.72, 'Underweight', 100.7],
 ['Bruno', 'Bulldog', 'Bad', 35.56, 'Normal', 104.1],
 ['Rosie', 'Beagle', 'Good', 33.02, 'Normal', 100.6],
 ['Duke', 'Doberman Pinscher', 'Good', 71.12, 'Normal', 100.1],
 ['Luna', 'Shiba Inu', 'Good', 43.18, 'Obese', 100.3],

#### 3) Convert the Good Doggy? column to a boolean. ####

In [4]:
for i, row in enumerate(dog_data):
    if i > 0:
        row[2] = (row[2] == 'Good') # returns true if good, otherwise false

dog_data

[['Name',
  'Breed',
  'Good Doggy?',
  'Length [cm]',
  'Weight Range',
  'Body Temp [F]'],
 ['Buddy', 'Golden Retriever', True, 60.96, 'Overweight', 101.5],
 ['Cooper', 'Doberman Pinscher', False, 71.12, 'Normal', 105.2],
 ['Max', 'German Shepherd', True, 66.04, 'Normal', 100.2],
 ['Bella', 'Chow Chow', False, 50.8, 'Normal', 100.5],
 ['Lucy', 'Yorkshire Terrier', True, 22.86, 'Normal', 102.0],
 ['Charlie', 'Beagle', True, 33.02, 'Overweight', 100.4],
 ['Daisy', 'Akita', False, 60.96, 'Normal', 99.8],
 ['Bailey', 'Jack Russell Terrier', False, 25.4, 'Underweight', 101.2],
 ['Lola', 'Rottweiler', True, 68.58, 'Normal', 102.5],
 ['Sadie', 'Siberian Husky', False, 58.42, 'Normal', 99.5],
 ['Penny', 'Standard Poodle', True, 45.72, 'Underweight', 100.7],
 ['Bruno', 'Bulldog', False, 35.56, 'Normal', 104.1],
 ['Rosie', 'Beagle', True, 33.02, 'Normal', 100.6],
 ['Duke', 'Doberman Pinscher', True, 71.12, 'Normal', 100.1],
 ['Luna', 'Shiba Inu', True, 43.18, 'Obese', 100.3],
 ['Scout', 'Golde

#### 4) Convert the Weight column to an integer. ####

In [5]:
value_mapper = {'Underweight': 0, 'Normal': 1, 'Overweight': 2, 'Obese': 3}

for i, row in enumerate(dog_data):
    if i > 0:
        row[4] = value_mapper[row[4]]

dog_data

[['Name',
  'Breed',
  'Good Doggy?',
  'Length [cm]',
  'Weight Range',
  'Body Temp [F]'],
 ['Buddy', 'Golden Retriever', True, 60.96, 2, 101.5],
 ['Cooper', 'Doberman Pinscher', False, 71.12, 1, 105.2],
 ['Max', 'German Shepherd', True, 66.04, 1, 100.2],
 ['Bella', 'Chow Chow', False, 50.8, 1, 100.5],
 ['Lucy', 'Yorkshire Terrier', True, 22.86, 1, 102.0],
 ['Charlie', 'Beagle', True, 33.02, 2, 100.4],
 ['Daisy', 'Akita', False, 60.96, 1, 99.8],
 ['Bailey', 'Jack Russell Terrier', False, 25.4, 0, 101.2],
 ['Lola', 'Rottweiler', True, 68.58, 1, 102.5],
 ['Sadie', 'Siberian Husky', False, 58.42, 1, 99.5],
 ['Penny', 'Standard Poodle', True, 45.72, 0, 100.7],
 ['Bruno', 'Bulldog', False, 35.56, 1, 104.1],
 ['Rosie', 'Beagle', True, 33.02, 1, 100.6],
 ['Duke', 'Doberman Pinscher', True, 71.12, 1, 100.1],
 ['Luna', 'Shiba Inu', True, 43.18, 3, 100.3],
 ['Scout', 'Golden Retriever', True, 55.88, 1, 103.5],
 ['Rex', 'Rottweiler', False, 58.42, 1, 100.2]]

#### 5) Create a new column named Has_Fever. Value should be 0 if temp is normal, 1 if temp is above 102 F. ####

In [6]:
for i, row in enumerate(dog_data):
    if i == 0:
        row.append("Has_Fever")
    else:
        row.append(row[5] > 102.0)

dog_data

[['Name',
  'Breed',
  'Good Doggy?',
  'Length [cm]',
  'Weight Range',
  'Body Temp [F]',
  'Has_Fever'],
 ['Buddy', 'Golden Retriever', True, 60.96, 2, 101.5, False],
 ['Cooper', 'Doberman Pinscher', False, 71.12, 1, 105.2, True],
 ['Max', 'German Shepherd', True, 66.04, 1, 100.2, False],
 ['Bella', 'Chow Chow', False, 50.8, 1, 100.5, False],
 ['Lucy', 'Yorkshire Terrier', True, 22.86, 1, 102.0, False],
 ['Charlie', 'Beagle', True, 33.02, 2, 100.4, False],
 ['Daisy', 'Akita', False, 60.96, 1, 99.8, False],
 ['Bailey', 'Jack Russell Terrier', False, 25.4, 0, 101.2, False],
 ['Lola', 'Rottweiler', True, 68.58, 1, 102.5, True],
 ['Sadie', 'Siberian Husky', False, 58.42, 1, 99.5, False],
 ['Penny', 'Standard Poodle', True, 45.72, 0, 100.7, False],
 ['Bruno', 'Bulldog', False, 35.56, 1, 104.1, True],
 ['Rosie', 'Beagle', True, 33.02, 1, 100.6, False],
 ['Duke', 'Doberman Pinscher', True, 71.12, 1, 100.1, False],
 ['Luna', 'Shiba Inu', True, 43.18, 3, 100.3, False],
 ['Scout', 'Golden Ret